adapter/src/debug_session.rs

```diff
@@ -1103,6 +1133,85 @@ impl DebugSession {
         Ok(())
     }
 
+    fn handle_disassemble(&mut self, args: DisassembleArguments) -> Result<DisassembleResponseBody, Error> {
+        fn invalid_instruction() -> DisassembledInstruction {
+            DisassembledInstruction {
+                address: "".into(),
+                instruction: "<invalid>".into(),
+                ..Default::default()
+            }
+        }
+
+        let base_addr = parse_int::parse::<u64>(&args.memory_reference)?;
+        let base_addr = match args.offset {
+            Some(offset) => base_addr.wrapping_add(offset as u64),
+            None => base_addr,
+        };
+        let instruction_offset = args.instruction_offset.unwrap_or(0);
+        if args.instruction_count < 0 {
+            bail!("Invalid instruction_count");
+        }
+        let instruction_count = args.instruction_count as usize;
+
+        let mut result = if instruction_offset >= 0 {
+            let start_addr = SBAddress::from_load_address(base_addr, &self.target);
+            let instructions =
+                self.target.read_instructions(&start_addr, (instruction_offset + args.instruction_count) as u32, &self.assembly_flavor);
+
+            let mut dis_instructions = Vec::new();
+            for instr in instructions.iter().skip(instruction_offset as usize) {
+                dis_instructions.push(disassembly::sbinstr_to_disinstr(&instr, &self.target, true));
+            }
+            dis_instructions
+        } else {
+            let bytes_per_instruction = disassembly::max_instruction_bytes(&self.target);
+            let offset_bytes = -instruction_offset * bytes_per_instruction as i64;
+            let start_addr = base_addr.wrapping_sub(offset_bytes as u64);
+            let mut disassemble_bytes = instruction_count * bytes_per_instruction as usize;
+
+            let mut dis_instructions = Vec::new();
+
+            let expected_index = -instruction_offset as usize;
+
+            // we make sure to extend disassemble_bytes to ensure that base_addr
+            // is always included
+            if start_addr + ( disassemble_bytes as u64 ) < base_addr {
+              disassemble_bytes = ( base_addr - start_addr + bytes_per_instruction ) as usize;
+            }
+
+            for shuffle_count in 0..bytes_per_instruction {
+              dis_instructions = disassembly::disassemble_byte_range(start_addr - shuffle_count, disassemble_bytes, &self.target.process(), &self.assembly_flavor)?;
+              if let Some(index) = dis_instructions.iter().position( |i| parse_int::parse::<u64>(&i.address).unwrap_or(0) == base_addr ) {
+                // we need to make sure that the entry for the requested
+                // address, is precicely at the index expected, i.e.
+                // -instruction_offset
+                if index < expected_index {
+                  // pad the start with expected_index - index dummy
```

Comment:
Hm, under what circumstances does this happen?

Reply:
Good question. I don't actually know, but it would be catastrophic for the [totally mad DAP protocol](https://github.com/microsoft/debug-adapter-protocol/issues/335#issuecomment-1269399532) if it did.

I'm not sure that I can craft an actual test case, but _maybe_ things like low addresses, or ones which end up (after offsetting) in kernel pages or something. Maybe we'd just get back NOPs, but I'm not sure. I felt more secure doing this based on the above protocol explanation than leaving it. Certainly after I did this refactoring and added this pre/post padding, everything started to work in both clients.

Reply:
I haven't and it _looks like_ lldb just fills in the blanks with NOPs. Taking the following 'tiny.c':

```c
int main(int argc, char**argv)
{
  int res = argc + 1;
  return res;
}
```

which compiles on macOS arm64 to:

```
ben@BenMBP2021 simple % objdump -D tiny

tiny:	file format mach-o arm64

Disassembly of section __TEXT,__text:

0000000100003f90 <_main>:
100003f90: ff 83 00 d1 	sub	sp, sp, #32
100003f94: ff 1f 00 b9 	str	wzr, [sp, #28]
100003f98: e0 1b 00 b9 	str	w0, [sp, #24]
100003f9c: e1 0b 00 f9 	str	x1, [sp, #16]
100003fa0: e8 1b 40 b9 	ldr	w8, [sp, #24]
100003fa4: 08 05 00 11 	add	w8, w8, #1
100003fa8: e8 0f 00 b9 	str	w8, [sp, #12]
100003fac: e0 0f 40 b9 	ldr	w0, [sp, #12]
100003fb0: ff 83 00 91 	add	sp, sp, #32
100003fb4: c0 03 5f d6 	ret

Disassembly of section __TEXT,__unwind_info:

0000000100003fb8 <__unwind_info>:
100003fb8: 01 00 00 00 	udf	#1
100003fbc: 1c 00 00 00 	udf	#28
100003fc0: 00 00 00 00 	udf	#0
100003fc4: 1c 00 00 00 	udf	#28
100003fc8: 00 00 00 00 	udf	#0
100003fcc: 1c 00 00 00 	udf	#28
100003fd0: 02 00 00 00 	udf	#2
100003fd4: 90 3f 00 00 	udf	#16272
100003fd8: 34 00 00 00 	udf	#52
100003fdc: 34 00 00 00 	udf	#52
100003fe0: b9 3f 00 00 	udf	#16313
100003fe4: 00 00 00 00 	udf	#0
100003fe8: 34 00 00 00 	udf	#52
100003fec: 03 00 00 00 	udf	#3
100003ff0: 0c 00 01 00 	<unknown>
100003ff4: 10 00 01 00 	<unknown>
100003ff8: 00 00 00 00 	udf	#0
100003ffc: 00 20 00 02 	<unknown>
```

we have the following output from codelldb using these patches. It looks like lldb just filled in `udf #0` (i.e. NOP I guess):


```
0x0000000100003eac:		udf #0x0
0x0000000100003eb0:		udf #0x0
0x0000000100003eb4:		udf #0x0
0x0000000100003eb8:		udf #0x0
0x0000000100003ebc:		udf #0x0
0x0000000100003ec0:		udf #0x0
0x0000000100003ec4:		udf #0x0
0x0000000100003ec8:		udf #0x0
0x0000000100003ecc:		udf #0x0
0x0000000100003ed0:		udf #0x0
0x0000000100003ed4:		udf #0x0
0x0000000100003ed8:		udf #0x0
0x0000000100003edc:		udf #0x0
0x0000000100003ee0:		udf #0x0
0x0000000100003ee4:		udf #0x0
0x0000000100003ee8:		udf #0x0
0x0000000100003eec:		udf #0x0
0x0000000100003ef0:		udf #0x0
0x0000000100003ef4:		udf #0x0
0x0000000100003ef8:		udf #0x0
0x0000000100003efc:		udf #0x0
0x0000000100003f00:		udf #0x0
0x0000000100003f04:		udf #0x0
0x0000000100003f08:		udf #0x0
0x0000000100003f0c:		udf #0x0
0x0000000100003f10:		udf #0x0
0x0000000100003f14:		udf #0x0
0x0000000100003f18:		udf #0x0
0x0000000100003f1c:		udf #0x0
0x0000000100003f20:		udf #0x0
0x0000000100003f24:		udf #0x0
0x0000000100003f28:		udf #0x0
0x0000000100003f2c:		udf #0x0
0x0000000100003f30:		udf #0x0
0x0000000100003f34:		udf #0x0
0x0000000100003f38:		udf #0x0
0x0000000100003f3c:		udf #0x0
0x0000000100003f40:		udf #0x0
0x0000000100003f44:		udf #0x0
0x0000000100003f48:		udf #0x0
0x0000000100003f4c:		udf #0x0
0x0000000100003f50:		udf #0x0
0x0000000100003f54:		udf #0x0
0x0000000100003f58:		udf #0x0
0x0000000100003f5c:		udf #0x0
0x0000000100003f60:		udf #0x0
0x0000000100003f64:		udf #0x0
0x0000000100003f68:		udf #0x0
0x0000000100003f6c:		udf #0x0
0x0000000100003f70:		udf #0x0
0x0000000100003f74:		udf #0x0
0x0000000100003f78:		udf #0x0
0x0000000100003f7c:		udf #0x0
0x0000000100003f80:		udf #0x0
0x0000000100003f84:		udf #0x0
0x0000000100003f88:		udf #0x0
0x0000000100003f8c:		udf #0x0
0x0000000100003f90:		sub sp, sp, #0x20
0x0000000100003f94:		str wzr, [sp, #0x1c]
0x0000000100003f98:		str w0, [sp, #0x18]
0x0000000100003f9c:		str x1, [sp, #0x10]
0x0000000100003fa0:		ldr w8, [sp, #0x18]
0x0000000100003fa4:		add w8, w8, #0x1
0x0000000100003fa8:		str w8, [sp, #0xc]
0x0000000100003fac:		ldr w0, [sp, #0xc]
0x0000000100003fb0:		add sp, sp, #0x20
0x0000000100003fb4:		ret 
0x0000000100003fb8:		udf #0x1
0x0000000100003fbc:		udf #0x1c
0x0000000100003fc0:		udf #0x0
0x0000000100003fc4:		udf #0x1c
0x0000000100003fc8:		udf #0x0
0x0000000100003fcc:		udf #0x1c
0x0000000100003fd0:		udf #0x2
0x0000000100003fd4:		udf #0x3f90
0x0000000100003fd8:		udf #0x34
0x0000000100003fdc:		udf #0x34
0x0000000100003fe0:		udf #0x3fb9
0x0000000100003fe4:		udf #0x0
0x0000000100003fe8:		udf #0x34
0x0000000100003fec:		udf #0x3
0x0000000100003ff0:		.long 0x0001000c; unknown opcode
0x0000000100003ff4:		.long 0x00010010; unknown opcode
0x0000000100003ff8:		udf #0x0
0x0000000100003ffc:		.long 0x02002000; unknown opcode
0x0000000100004000:		udf #0x0
0x0000000100004004:		udf #0x20
0x0000000100004008:		udf #0x30
0x000000010000400c:		udf #0x30
0x0000000100004010:		udf #0x0
0x0000000100004014:		udf #0x1
0x0000000100004018:		udf #0x0
0x000000010000401c:		udf #0x0
0x0000000100004020:		udf #0x3
0x0000000100004024:		udf #0x0
0x0000000100004028:		udf #0x0
0x000000010000402c:		udf #0x0
0x0000000100004030:		udf #0x0
0x0000000100004034:		udf #0x0
0x0000000100004038:		.long 0x005f0100; unknown opcode
0x000000010000403c:		udf #0x209
0x0000000100004040:		.long 0x5f020000; unknown opcode
0x0000000100004044:		fcmeq p13.h, p2/z, z3.h, z31.h
0x0000000100004048:		.long 0x75636578; unknown opcode
0x000000010000404c:		.long 0x685f6574; unknown opcode
0x0000000100004050:		fnmls z5.h, p0/m, z11.h, z4.h
0x0000000100004054:		stp d18, d0, [x3, #0x50]
0x0000000100004058:		.long 0x006e6961; unknown opcode
0x000000010000405c:		adrp x5, 100
0x0000000100004060:		udf #0x7f
0x0000000100004064:		udf #0x0
0x0000000100004068:		udf #0x7f90
0x000000010000406c:		udf #0x0
0x0000000100004070:		udf #0x1
0x0000000100004074:		udf #0x164
0x0000000100004078:		udf #0x0
0x000000010000407c:		udf #0x0
0x0000000100004080:		udf #0x1c
0x0000000100004084:		udf #0x64
0x0000000100004088:		udf #0x0
0x000000010000408c:		udf #0x0
0x0000000100004090:		udf #0x59
```

You can see the main function in the middle there and the junk surrounding it.

We might get away with removing this code block, but part of me thinks itself documents the required behaviour, even if lldb's behaviour changes (e.g. on another platform)

Reply:
Interestingly, on x86 we get slightly different:

```
0x0000000100003f28:	00 00                	addb   %al, (%rax)
0x0000000100003f2a:	00 00                	addb   %al, (%rax)
0x0000000100003f2c:	00 00                	addb   %al, (%rax)
0x0000000100003f2e:	00 00                	addb   %al, (%rax)
0x0000000100003f30:	00 00                	addb   %al, (%rax)
0x0000000100003f32:	00 00                	addb   %al, (%rax)
0x0000000100003f34:	00 00                	addb   %al, (%rax)
0x0000000100003f36:	00 00                	addb   %al, (%rax)
0x0000000100003f38:	00 00                	addb   %al, (%rax)
0x0000000100003f3a:	00 00                	addb   %al, (%rax)
0x0000000100003f3c:	00 00                	addb   %al, (%rax)
0x0000000100003f3e:	00 00                	addb   %al, (%rax)
0x0000000100003f40:	00 00                	addb   %al, (%rax)
0x0000000100003f42:	00 00                	addb   %al, (%rax)
0x0000000100003f44:	00 00                	addb   %al, (%rax)
0x0000000100003f46:	00 00                	addb   %al, (%rax)
0x0000000100003f48:	00 00                	addb   %al, (%rax)
0x0000000100003f4a:	00 00                	addb   %al, (%rax)
0x0000000100003f4c:	00 00                	addb   %al, (%rax)
0x0000000100003f4e:	00 00                	addb   %al, (%rax)
0x0000000100003f50:	00 00                	addb   %al, (%rax)
0x0000000100003f52:	00 00                	addb   %al, (%rax)
0x0000000100003f54:	00 00                	addb   %al, (%rax)
0x0000000100003f56:	00 00                	addb   %al, (%rax)
0x0000000100003f58:	00 00                	addb   %al, (%rax)
0x0000000100003f5a:	00 00                	addb   %al, (%rax)
0x0000000100003f5c:	00 00                	addb   %al, (%rax)
0x0000000100003f5e:	00 00                	addb   %al, (%rax)
0x0000000100003f60:	00 00                	addb   %al, (%rax)
0x0000000100003f62:	00 00                	addb   %al, (%rax)
0x0000000100003f64:	00 00                	addb   %al, (%rax)
0x0000000100003f66:	00 00                	addb   %al, (%rax)
0x0000000100003f68:	00 00                	addb   %al, (%rax)
0x0000000100003f6a:	00 00                	addb   %al, (%rax)
0x0000000100003f6c:	00 00                	addb   %al, (%rax)
0x0000000100003f6e:	00 00                	addb   %al, (%rax)
0x0000000100003f70:	00 00                	addb   %al, (%rax)
0x0000000100003f72:	00 00                	addb   %al, (%rax)
0x0000000100003f74:	00 00                	addb   %al, (%rax)
0x0000000100003f76:	00 00                	addb   %al, (%rax)
0x0000000100003f78:	00 00                	addb   %al, (%rax)
0x0000000100003f7a:	00 00                	addb   %al, (%rax)
0x0000000100003f7c:	00 00                	addb   %al, (%rax)
0x0000000100003f7e:	00 00                	addb   %al, (%rax)
0x0000000100003f80:	00 00                	addb   %al, (%rax)
0x0000000100003f82:	00 00                	addb   %al, (%rax)
0x0000000100003f84:	00 00                	addb   %al, (%rax)
0x0000000100003f86:	00 00                	addb   %al, (%rax)
0x0000000100003f88:	00 00                	addb   %al, (%rax)
0x0000000100003f8a:	00 00                	addb   %al, (%rax)
0x0000000100003f8c:	00 00                	addb   %al, (%rax)
0x0000000100003f8e:	00 00                	addb   %al, (%rax)
0x0000000100003f90:	55                   	pushq  %rbp
0x0000000100003f91:	48 89 E5             	movq   %rsp, %rbp
0x0000000100003f94:	C7 45 FC 00 00 00 00 	movl   $0x0, -0x4(%rbp)
0x0000000100003f9b:	89 7D F8             	movl   %edi, -0x8(%rbp)
0x0000000100003f9e:	48 89 75 F0          	movq   %rsi, -0x10(%rbp)
0x0000000100003fa2:	8B 45 F8             	movl   -0x8(%rbp), %eax
0x0000000100003fa5:	83 C0 01             	addl   $0x1, %eax
0x0000000100003fa8:	89 45 EC             	movl   %eax, -0x14(%rbp)
0x0000000100003fab:	8B 45 EC             	movl   -0x14(%rbp), %eax
0x0000000100003fae:	5D                   	popq   %rbp
0x0000000100003faf:	C3                   	retq   
0x0000000100003fb0:	01 00                	addl   %eax, (%rax)
0x0000000100003fb2:	00 00                	addb   %al, (%rax)
0x0000000100003fb4:	1C 00                	sbbb   $0x0, %al
0x0000000100003fb6:	00 00                	addb   %al, (%rax)
0x0000000100003fb8:	00 00                	addb   %al, (%rax)
0x0000000100003fba:	00 00                	addb   %al, (%rax)
0x0000000100003fbc:	1C 00                	sbbb   $0x0, %al
0x0000000100003fbe:	00 00                	addb   %al, (%rax)
0x0000000100003fc0:	00 00                	addb   %al, (%rax)
0x0000000100003fc2:	00 00                	addb   %al, (%rax)
0x0000000100003fc4:	1C 00                	sbbb   $0x0, %al
0x0000000100003fc6:	00 00                	addb   %al, (%rax)
0x0000000100003fc8:	02 00                	addb   (%rax), %al
0x0000000100003fca:	00 00                	addb   %al, (%rax)
0x0000000100003fcc:	90                   	nop    
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
0x0000000000000000:	                     	<invalid>
```